chore(deadline): add secrets management documentation (#558)
aws/aws-rfdk: packages/aws-rfdk/docs/upgrade/upgrading-0.38.md

# Upgrading to RFDK v0.38.x or Newer

Starting in RFDK v0.38.0, [Deadline Secrets Management](https://docs.thinkboxsoftware.com/products/deadline/10.1/1_User%20Manual/manual/secrets-management/deadline-secrets-management.html) will be enabled on the

> `Repository` and `RenderQueue` constructs by default when using Deadline 10.1.18 and later. This can cause some issues with existing farm configurations because Deadline Secrets Management requires:

Reviewer: ```suggestion
`Repository` and `RenderQueue` constructs by default when using Deadline 10.1.19 and later. This can cause some issues with existing farm configurations because Deadline Secrets Management requires:
```

Author: Updated

- The internal protocol on the Render Queue to be HTTPS
- The external protocol on the Render Queue to be HTTPS

> Additionally, if you plan on upgrading to RFDK 0.38 or newer and Deadline 10.1.18 or newer, but do **not** want to enable Deadline Secrets Management, you may have to make changes to your RFDK application so that

Reviewer: ```suggestion
Additionally, if you plan on upgrading to RFDK 0.38 or newer and Deadline 10.1.19 or newer, but do **not** want to enable Deadline Secrets Management, you may have to make changes to your RFDK application so that
```

Author: Updated

RFDK does not automatically configure Deadline Secrets Management on your farm.

> ## Upgrading Farms and Enabling Deadline Secrets Management

Reviewer: nit: can this be renamed to "Upgrading RFDK and Enabling ..."?

same feedback for the next section (" Upgrading Farms Without Enabling Deadline Secrets Management")

Author: Sure, updated

---

_This section describes how to upgrade your RFDK farm while also enabling Deadline Secrets Management. If you would **not** like to enable Deadline Secrets Management, please skip to the
next section "Upgrading Farms Without Enabling Deadline Secrets Management"._

---

To determine whether your `RenderQueue` is using HTTPS for internal and external protocols, you need to check what values you have set
for the [`trafficEncryption`](https://docs.aws.amazon.com/rfdk/api/latest/docs/aws-rfdk.deadline.RenderQueue.html#trafficencryption) property.
Your `RenderQueue` is using HTTPS for internal and external protocols if one of the following is true:

1. You have not specified the `trafficEncryption` or the `trafficEncryption.externalTLS` property for the `RenderQueue`
2. You have specified the `trafficEncryption` property for the `RenderQueue` and the following are true:
    1. The `internalProtocol` is `ApplicationProtocol.HTTPS`
    2. One of the following is true for the `externalTLS` property:
        1. `enabled` is `true`
        2. There are values for both `acmCertificate` and `acmCertificateChain`
        3. There is a value for `rfdkCertificate`

> If the above is true, all you need to do to enable Deadline Secrets Management is redeploy your RFDK application.

Reviewer: This is still a little unclear to me. Allow me to propose an alternative here:

---

Your `RenderQueue` will use HTTPS for both internal and external protocols for RFDK 0.38.0 if **EITHER of the following are true**:

1.  You **HAVE NOT** specified the `trafficEncryption`  (TypeScript) / `traffic_encryption` (Python) property
2.  You **HAVE** specified the `trafficEncryption`  (TypeScript) / `traffic_encryption` (Python) property **AND BOTH of the following are true**:
    1.  **EITHER** of the following are true:
        1.  You **HAVE NOT** specified the `trafficEncryption.internalProtocol` (TypeScript) / `traffic_encryption.internal_protocol` (Python)
        2.  You **HAVE** specified the `trafficEncryption.internalProtocol` (TypeScript) / `traffic_encryption.internal_protocol` (Python) and it is set to `ApplicationProtocol.HTTPS`
    2.  **EITHER** of the following are true:
        1.  You **HAVE NOT** specified the `trafficEncryption.externalTLS`  (TypeScript) / `traffic_encryption.external_tls` (Python)
        2.  You **HAVE NOT** specified `trafficEncryption.externalTLS.enabled` (TypeScript) / `traffic_encryption.external_tls.enabled` (Python)
        3.  You **HAVE** specified `trafficEncryption.externalTLS.enabled` (TypeScript) / `traffic_encryption.external_tls.enabled` (Python) and it is set to `true` (TypeScript) / `True` (Python)

If you have determined that your `RenderQueue` will use HTTPS for both internal and external protocols in RFDK 0.38.0, then all you need to do to enable Deadline Secrets Management is to upgrade you RFDK application to use 0.30.0 and re-deploy.

Author: Sounds good, updated

If your `RenderQueue` is not using HTTPS for its internal and/or external protocols as described above, you will need to satisfy the conditions above before deploying your farm so that
RFDK can successfully configure Deadline Secrets Management for you.

The below example code snippets demonstrate what each case above may look like:

<details><summary><b>Typescript</b> (click to expand)</summary>

```ts
import { ApplicationProtocol } from '@aws-cdk/aws-elasticloadbalancingv2';
import { RenderQueue } from 'aws-rfdk/deadline';

new RenderQueue(this, 'RenderQueue', {

>   // no "trafficEncryption" property, so TLS is enabled externally by default

Reviewer: ```suggestion
  // no "trafficEncryption" property, so TLS is enabled both internally and externally by default
```

Author: Updated in both languages

  // ...
});

> // OR

Reviewer: nit: a few edge cases missing here:

```ts
new RenderQueue(this, 'RenderQueue', {
  // ...
  trafficEncryption: {
    // No "internalProtocol" property, so TLS is enabled internally by default
    // No "externalTLS" property, so TLS is enabled externally by default
  },
  // ...
});
```

```ts
new RenderQueue(this, 'RenderQueue', {
  // ...
  trafficEncryption: {
    // No "internalProtocol" property, so TLS is enabled internally by default
    externalTLS: {
      // No "enabled" property, so external TLS will be enabled by default
    }
  },
  // ...
});
```

Author: Added these in for both languages

  },
  // ...
});

// OR

new RenderQueue(this, 'RenderQueue', {
  // ...
  trafficEncryption: {
    internalProtocol: ApplicationProtocol.HTTPS,
    externalTLS: {
      enabled: true,
    },
  },
  // ...
});

// OR

const certificate = // ...your ACM certificate
const certificateChain = // ...your ACM certificate chain
new RenderQueue(this, 'RenderQueue', {
  // ...
  trafficEncryption: {
    internalProtocol: ApplicationProtocol.HTTPS,
    externalTLS: {
      acmCertificate: certificate,
      acmCertificateChain: certificateChain,
    },
  },
  // ...
});

// OR

const certificate = // ...your X509 RFDK certificate
new RenderQueue(this, 'RenderQueue', {
  // ...
  trafficEncryption: {
    internalProtocol: ApplicationProtocol.HTTPS,
    externalTLS: {
      rfdkCertificate: certificate,
    },
  },
  // ...
});
```

</details>

<details><summary><b>Python</b> (click to expand)</summary>

```python
from aws_cdk.aws_elasticloadbalancingv2 import ApplicationProtocol
from aws_rfdk.deadline import (
    RenderQueue,
    RenderQueueExternalTLSProps,
    RenderQueueTrafficEncryptionProps,
)

RenderQueue(self, 'RenderQueue',

>   # no "traffic_encryption" property, so TLS is enabled externally by default

Reviewer: ```suggestion
  # no "traffic_encryption" property, so TLS is enabled both internally and externally by default
```

Author: Updated

# Upgrading to RFDK v0.38.x or Newer

Starting in RFDK v0.38.0, [Deadline Secrets Management](https://docs.thinkboxsoftware.com/products/deadline/10.1/1_User%20Manual/manual/secrets-management/deadline-secrets-management.html) will be enabled on the
`Repository` and `RenderQueue` constructs by default when using Deadline 10.1.19 and later. This can cause some issues with existing farm configurations because Deadline Secrets Management requires:

> - The internal protocol on the Render Queue to be HTTPS
> - The external protocol on the Render Queue to be HTTPS

Reviewer 2: ```suggestion
- The internal protocol on the Render Queue to be HTTPS. Internal here refers to traffic between the Remote Connection Server and load balancer).
- The external protocol on the Render Queue to be HTTPS. External here refers to traffic between the load balancer and clients.
```

Author: Updated

Additionally, if you plan on upgrading to RFDK 0.38 or newer and Deadline 10.1.19 or newer, but do **not** want to enable Deadline Secrets Management, you may have to make changes to your RFDK application so that
RFDK does not automatically configure Deadline Secrets Management on your farm.

## Upgrading RFDK and Enabling Deadline Secrets Management

---

_This section describes how to upgrade your RFDK farm while also enabling Deadline Secrets Management. If you would **not** like to enable Deadline Secrets Management, please skip to the

> next section "Upgrading Farms Without Enabling Deadline Secrets Management"._

Reviewer: ```suggestion
next section "Upgrading RFDK Without Enabling Deadline Secrets Management"._
```

Author: Thanks for catching that, updated

---

Your `RenderQueue` will use HTTPS for both internal and external protocols for RFDK 0.38.0 if **EITHER** of the following are true:

1. You **HAVE NOT** specified the `trafficEncryption` (TypeScript) / `traffic_encryption` (Python) property
1. You **HAVE** specified the `trafficEncryption` (TypeScript) / `traffic_encryption` (Python) property **AND BOTH of the following are true**:
    1. **EITHER** of the following are true:
        1. You **HAVE NOT** specified the `trafficEncryption.internalProtocol` (TypeScript) / `traffic_encryption.internal_protocol` (Python)
        1. You **HAVE** specified the `trafficEncryption.internalProtocol` (TypeScript) / `traffic_encryption.internal_protocol` (Python) and it is set to `ApplicationProtocol.HTTPS`
    1. **EITHER** of the following are true:
        1. You **HAVE NOT** specified the `trafficEncryption.externalTLS` (TypeScript) / `traffic_encryption.external_tls` (Python)
        1. You **HAVE NOT** specified `trafficEncryption.externalTLS.enabled` (TypeScript) / `traffic_encryption.external_tls.enabled` (Python)

>         1. You **HAVE** specified `trafficEncryption.externalTLS.enabled` (TypeScript) / `traffic_encryption.external_tls.enabled` (Python) and it is set to `true` (TypeScript) / `True` (Python)

Reviewer 2: This seems really wordy way of explaining that HTTPS is on by default or uses the configured value. It also doesn't explain why we care about HTTPS; it does get explained above, but I feel like we should re-iterate it since this is a different section. Can't we sum it up like this:

```
The first step to making sure we're ready to enable Secrets Management is to make sure we have HTTPS 
enabled on the `RenderQueue` for both the internal and external protocols. If you are upgrading from a version 
prior to 0.37, you should refer to the [v0.37 upgrade doc](./upgrading-0.37.md) for details about how the 
`RenderQueue` will default to using HTTPS now.

If any of the optional fields for configuring the protocol the `RenderQueue` uses aren't set, it will default to using HTTPS. 
* Unless `trafficEncryption.internalProtocol` is set to `ApplicationProtocol.HTTP`, the internal protocol will be HTTPS.
* Unless `trafficEncryption.externalTLS.enabled` is set to `false`, the external protocol will be HTTPS.
```

Reviewer 2: Let's leave what you currently have as-is, but I still think adding the first little blurb I wrote out would be helpful:

```
The first step to making sure we're ready to enable Secrets Management is to make sure we have HTTPS 
enabled on the `RenderQueue` for both the internal and external protocols. If you are upgrading from a version 
prior to 0.37, you should refer to the [v0.37 upgrade doc](./upgrading-0.37.md) for details about how the 
`RenderQueue` will default to using HTTPS now.
```

Author: Sounds good, added

> The below example code snippets demonstrate what each case above may look like:

Reviewer 2: If we go with my suggestion above we'd likely want to change this as well:
```suggestion
The below example code snippets demonstrate all the possible ways for HTTPS to be enabled:
```

Reviewer 2: Let's skip this suggestion

> your RFDK application to use 0.38.0 and re-deploy.

Reviewer: I know the logic is not yet implemented, but we plan to change this to reflect our plan to have secrets management only enabled when using Deadline version 10.1.19 or higher. What do you think about changing these docs here in advance?

```suggestion
your RFDK application to use RFDK `0.38.x` and Deadline `10.1.19.x` or higher and re-deploy.
```

Also, does it make sense to move this paragraph below the example snippets? What do you think?

Author: Sounds good, and yeah it flows better when it's below the examples. Updated

  ),
  # ...
)

# OR

RenderQueue(self, 'RenderQueue',
  # ...
  traffic_encryption=RenderQueueTrafficEncryptionProps(
    internal_protocol=ApplicationProtocol.HTTPS,
    external_tls=RenderQueueExternalTLSProps(
      enabled=True,
    ),
  ),
  # ...
)

# OR

certificate = # ...your ACM certificate
certificate_chain = # ...your ACM certificate chain
RenderQueue(self, 'RenderQueue',
  # ...
  traffic_encryption=RenderQueueTrafficEncryptionProps(
    internal_protocol=ApplicationProtocol.HTTPS,
    external_tls=RenderQueueExternalTLSProps(
      acmCertificate=certificate,
      acmCertificateChain=certificate_chain,
    ),
  ),
  # ...
)

# OR

certificate = # ...your X509 RFDK certificate
RenderQueue(self, 'RenderQueue',
  # ...
  traffic_encryption=RenderQueueTrafficEncryptionProps(
    internal_protocol=ApplicationProtocol.HTTPS,
    external_tls=RenderQueueExternalTLSProps(
      rfdk_certificate=certificate,
    ),
  ),
  # ...
)
```

</details>

If you have determined that your `RenderQueue` will use HTTPS for both internal and external protocols in RFDK 0.38.0, then all you need to do to enable Deadline Secrets Management is to upgrade

> your RFDK application to use RFDK `0.38.x` and Deadline `10.1.19.x` or higher and re-deploy.

Reviewer 2: What if you determine your Render Queue isn't using HTTPS? We should refer to the 0.37 upgrade doc for more detailed instructions about what will change if HTTPS is turned on.

Author: This has been added in the blurb from the suggestion above, so I think that should cover linking back to 0.37 upgrade docs and we don't need to add anything here. How does that sound?

Reviewer 2: That should be fine, but reading through this again, we don't really touch on the other things to consider when enabling Secrets Management, like dedicated subnets and how to set up credentials or where to find the ones that are generated. Maybe let's add a paragraph in about that.
```suggestion
If you have determined that your `RenderQueue` will use HTTPS for both internal and external protocols in RFDK 0.38.0, 
then the next step is to decide whether you'd like to set up your own credentials used to perform admin functions for 
Secrets Management or let RFDK generate them for you, and whether you should change your VPC to use dedicated 
subnets for different constructs. More details on these considerations can be found in the 
[RFDK Deadline readme](../../lib/deadline/README.md).

Finally, all you need to do to enable Deadline Secrets Management is to upgrade
your RFDK application to use RFDK `0.38.x` and Deadline `10.1.19.x` or higher and re-deploy.
```

Author: Good point, added in a section with that

## Upgrading Farms and Enabling Deadline Secrets Management

---

_This section describes how to upgrade your RFDK farm while also enabling Deadline Secrets Management. If you would **not** like to enable Deadline Secrets Management, please skip to the
next section "Upgrading Farms Without Enabling Deadline Secrets Management"._

---

To determine whether your `RenderQueue` is using HTTPS for internal and external protocols, you need to check what values you have set
for the [`trafficEncryption`](https://docs.aws.amazon.com/rfdk/api/latest/docs/aws-rfdk.deadline.RenderQueue.html#trafficencryption) property.
Your `RenderQueue` is using HTTPS for internal and external protocols if one of the following is true:

1. You have not specified the `trafficEncryption` property for the `RenderQueue`
2. You have specified the `trafficEncryption` property for the `RenderQueue` and the following are true:
    1. The `internalProtocol` is `ApplicationProtocol.HTTPS`
    2. One of the following is true for the `externalTLS` property:
        1. `enabled` is `true`
        2. There are values for both `acmCertificate` and `acmCertificateChain`

>         3. There is a value for `rfdkCertificate`

Reviewer: I believe `trafficEncryption.externalTLS` is optional and if not specified, it actually defaults to external HTTPS .

On a broader note, perhaps it might help to provide some example code snippets to illustrate the different setups like we have for the [0.27 upgrade docs](https://github.com/aws/aws-rfdk/blob/mainline/packages/aws-rfdk/docs/upgrade/upgrading-0.27.md). For example:

---

**TypeScript**:

```ts
import { ApplicationProtocol } from '@aws-cdk/aws-elasticloadbalancingv2';
import { RenderQueue,  } from 'aws-rfdk/deadline';

new RenderQueue(this, 'RenderQueue', {
  // ...
  trafficEncryption: {
    // Internal protocol is HTTPS
    internalProtocol: ApplicationProtocol.HTTPS,

    // No "externalTLS" property, so TLS is enabled externally by default
  },
  // ...
});

// OR

new RenderQueue(this, 'RenderQueue', {
  // ...
});

// OR

// ...
```

**Python**:

```py
from aws_cdk.aws_elasticloadbalancingv2 import ApplicationProtocol
from aws_rfdk.deadline import (
    RenderQueue,
    RenderQueueTrafficEncryptionProps
)

RenderQueue(self, 'RenderQueue',
  # ...
  traffic_encryption=RenderQueueTrafficEncryptionProps(
    # Internal protocol is HTTPS
    internalProtocol=ApplicationProtocol.HTTPS,

    # No "externalTLS" property, so TLS is enabled externally by default
  ),
  # ...
);

# OR

RenderQueue(this, 'RenderQueue',
  // ...
});

# OR

# ...
```

What do you think?

Author: Right, thanks for catching that, and having code samples sounds like a good idea, added. I used some formatting to make them expandable items since the code blocks are quite long, what do you think of it?
